# Project: Wrangling and Analyze Data

## Import Libraries 
In the cell below, we import **all** the libraries required for this project.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import requests
import seaborn as sns
import datetime

## Data Gathering
In the cell below, gather **all** three pieces of data for this project and load them in the notebook. **Note:** the methods required to gather each data are different.
1. Directly download the WeRateDogs Twitter archive data (twitter_archive_enhanced.csv)

In [2]:
twitter_archive_df = pd.read_csv('twitter-archive-enhanced.csv') # Direct download of the WeRateDogs Twitter archive data, and store inside a pandas dataframe

2. Use the Requests library to download the tweet image prediction (image_predictions.tsv)

In [3]:
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
image_predictions_response = requests.get(url, allow_redirects = True)

with open('image-predictions.tsv', mode='wb') as file:
  file.write(image_predictions_response.content)

image_predictions_df = pd.read_csv('image-predictions.tsv', sep='\t')

3. Use the Tweepy library to query additional data via the Twitter API (tweet_json.txt)

In [4]:
# Since my Twitter Developer account approval is pending, 
# I will be reading the data directly from the tweet-json.json file
tweet_data = pd.read_json('tweet-json.json', lines=True)
tweet_data = tweet_data.filter(['id', 'favorite_count', 'retweet_count'])
tweet_data = tweet_data.rename(columns={"id":"tweet_id"})
tweet_data.to_csv('tweet_data.csv')

## Assessing Data
In this section, detect and document at least **eight (8) quality issues and two (2) tidiness issue**. You must use **both** visual assessment
programmatic assessement to assess the data.

**Note:** pay attention to the following key points when you access the data.

* You only want original ratings (no retweets) that have images. Though there are 5000+ tweets in the dataset, not all are dog ratings and some are retweets.
* Assessing and cleaning the entire dataset completely would require a lot of time, and is not necessary to practice and demonstrate your skills in data wrangling. Therefore, the requirements of this project are only to assess and clean at least 8 quality issues and at least 2 tidiness issues in this dataset.
* The fact that the rating numerators are greater than the denominators does not need to be cleaned. This [unique rating system](http://knowyourmeme.com/memes/theyre-good-dogs-brent) is a big part of the popularity of WeRateDogs.
* You do not need to gather the tweets beyond August 1st, 2017. You can, but note that you won't be able to gather the image predictions for these tweets since you don't have access to the algorithm used.



### i. Accessing the twitter archive data

In [5]:
twitter_archive_df.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [6]:
# # Number of retweeted tweets
# len(twitter_archive_df[twitter_archive_df['retweeted_status_id'].notnull()])

In [7]:
twitter_archive_df.sample(10)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2207  668627278264475648                    NaN                  NaN   
810   771380798096281600                    NaN                  NaN   
2151  669682095984410625                    NaN                  NaN   
387   826598799820865537           8.265984e+17         4.196984e+09   
2159  669571471778410496                    NaN                  NaN   
1994  672604026190569472                    NaN                  NaN   
117   870063196459192321                    NaN                  NaN   
2109  670449342516494336                    NaN                  NaN   
1928  674045139690631169                    NaN                  NaN   
1736  679722016581222400                    NaN                  NaN   

                      timestamp  \
2207  2015-11-23 03:09:00 +0000   
810   2016-09-01 16:14:48 +0000   
2151  2015-11-26 01:00:28 +0000   
387   2017-02-01 01:11:25 +0000   
2159  2015-11-25 17:40:53 +0000   
1994  2015-12-04 02:31:10 +0000   
117   2017-05-31 23:43:25 +0000   
2109  2015-11-28 03:49:14 +0000   
1928  2015-12-08 01:57:39 +0000   
1736  2015-12-23 17:55:32 +0000   

                                                 source  \
2207  <a href="http://twitter.com/download/iphone" r...   
810   <a href="http://twitter.com/download/iphone" r...   
2151  <a href="http://twitter.com/download/iphone" r...   
387   <a href="http://twitter.com/download/iphone" r...   
2159  <a href="http://twitter.com/download/iphone" r...   
1994  <a href="http://twitter.com/download/iphone" r...   
117   <a href="http://twitter.com/download/iphone" r...   
2109  <a href="http://twitter.com/download/iphone" r...   
1928  <a href="http://twitter.com/download/iphone" r...   
1736  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
2207  This is Timofy. He's a pilot for Southwest. It...                  NaN   
810   Meet Fizz. She thinks love is a social constru...                  NaN   
2151  Meet Louie. He just pounded that bottle of win...                  NaN   
387   I was going to do 007/10, but the joke wasn't ...                  NaN   
2159  This is Keith. He's had 13 DUIs. 7/10 that's t...                  NaN   
1994  This is a baby Rand Paul. Curls for days. 11/1...                  NaN   
117   Meet Clifford. He's quite large. Also red. Goo...                  NaN   
2109  Vibrant dog here. Fabulous tail. Only 2 legs t...                  NaN   
1928  Herd of wild dogs here. Not sure what they're ...                  NaN   
1736  This is Mike. He is a Jordanian Frito Pilates....                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
2207                       NaN                        NaN   
810                        NaN                        NaN   
2151                       NaN                        NaN   
387                        NaN                        NaN   
2159                       NaN                        NaN   
1994                       NaN                        NaN   
117                        NaN                        NaN   
2109                       NaN                        NaN   
1928                       NaN                        NaN   
1736                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
2207  https://twitter.com/dog_rates/status/668627278...                 9   
810   https://twitter.com/dog_rates/status/771380798...                11   
2151  https://twitter.com/dog_rates/status/669682095...                 9   
387                                                 NaN                 7   
2159  https://twitter.com/dog_rates/status/669571471...                 7   
1994  https://twitter.com/dog_rates/status/672604026...                11   
117   https://twitter.com/dog_rates/status/870063196...                14   
2109  https://twit

In [8]:
twitter_archive_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [9]:
twitter_archive_df.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [10]:
twitter_archive_df['name'].value_counts()

None          745
a              55
Charlie        12
Cooper         11
Lucy           11
             ... 
Dex             1
Ace             1
Tayzie          1
Grizzie         1
Christoper      1
Name: name, Length: 957, dtype: int64

In [11]:
# Showing rows with numerators less than 10
twitter_archive_df[twitter_archive_df.rating_numerator < 10].count()[0]

440

In [12]:
# Showing rows with denominators equals 0
twitter_archive_df[(twitter_archive_df.rating_denominator == 0) ]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
313  835246439529840640           8.352460e+17           26259576.0   

                     timestamp  \
313  2017-02-24 21:54:03 +0000   

                                                source  \
313  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
313  @jonnysun @Lin_Manuel ok jomny I know you're e...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp expanded_urls  \
313                       NaN                        NaN           NaN   

     rating_numerator  rating_denominator  name doggo floofer pupper puppo  
313               960                   0  None  None    None   None  None

In [13]:
# Showing rows with denominators greater or less than 10
print(len(twitter_archive_df[(twitter_archive_df.rating_denominator < 10) | (twitter_archive_df.rating_denominator > 10)]))
twitter_archive_df[(twitter_archive_df.rating_denominator < 10) | (twitter_archive_df.rating_denominator > 10)]

23


tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
313   835246439529840640           8.352460e+17         2.625958e+07   
342   832088576586297345           8.320875e+17         3.058208e+07   
433   820690176645140481                    NaN                  NaN   
516   810984652412424192                    NaN                  NaN   
784   775096608509886464                    NaN                  NaN   
902   758467244762497024                    NaN                  NaN   
1068  740373189193256964                    NaN                  NaN   
1120  731156023742988288                    NaN                  NaN   
1165  722974582966214656                    NaN                  NaN   
1202  716439118184652801                    NaN                  NaN   
1228  713900603437621249                    NaN                  NaN   
1254  710658690886586372                    NaN                  NaN   
1274  709198395643068416                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
1433  697463031882764288                    NaN                  NaN   
1598  686035780142297088           6.860340e+17         4.196984e+09   
1634  684225744407494656           6.842229e+17         4.196984e+09   
1635  684222868335505415                    NaN                  NaN   
1662  682962037429899265                    NaN                  NaN   
1663  682808988178739200           6.827884e+17         4.196984e+09   
1779  677716515794329600                    NaN                  NaN   
1843  675853064436391936                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   

                      timestamp  \
313   2017-02-24 21:54:03 +0000   
342   2017-02-16 04:45:50 +0000   
433   2017-01-15 17:52:40 +0000   
516   2016-12-19 23:06:23 +0000   
784   2016-09-11 22:20:06 +0000   
902   2016-07-28 01:00:57 +0000   
1068  2016-06-08 02:41:38 +0000   
1120  2016-05-13 16:15:54 +0000   
1165  2016-04-21 02:25:47 +0000   
1202  2016-04-03 01:36:11 +0000   
1228  2016-03-27 01:29:02 +0000   
1254  2016-03-18 02:46:49 +0000   
1274  2016-03-14 02:04:08 +0000   
1351  2016-02-28 21:25:30 +0000   
1433  2016-02-10 16:51:59 +0000   
1598  2016-01-10 04:04:10 +0000   
1634  2016-01-05 04:11:44 +0000   
1635  2016-01-05 04:00:18 +0000   
1662  2016-01-01 16:30:13 +0000   
1663  2016-01-01 06:22:03 +0000   
1779  2015-12-18 05:06:23 +0000   
1843  2015-12-13 01:41:41 +0000   
2335  2015-11-16 16:11:11 +0000   

                                                 source  \
313   <a href="http://twitter.com/download/iphone" r...   
342   <a href="http://twitter.com/download/iphone" r...   
433   <a href="http://twitter.com/download/iphone" r...   
516   <a href="http://twitter.com/download/iphone" r...   
784   <a href="http://twitter.com/download/iphone" r...   
902   <a href="http://twitter.com/download/iphone" r...   
1068  <a href="http://twitter.com/download/iphone" r...   
1120  <a href="http://twitter.com/download/iphone" r...   
1165  <a href="http://twitter.com/download/iphone" r...   
1202  <a href="http://twitter.com/download/iphone" r...   
1228  <a href="http://twitter.com/download/iphone" r...   
1254  <a href="http://twitter.com/download/iphone" r...   
1274  <a href="http://twitter.com/download/iphone" r...   
1351  <a href="http://twitter.com/download/iphone" r...   
1433  <a href="http://twitter.com/download/iphone" r...   
1598  <a href="http://twitter.com/download/iphone" r...   
1634  <a href="http://twitter.com/download/iphone" r...   
1635  <a href="http://twitter.com/download/iphone" r...   
1662  <a href="http://twitter.com/download/iphone" r...   
1663  <a href="http://twitter.com/download/iphone" r...   
1779  <a href="http://twitter.com/download/iphone" r...   
1843  <a href="http://twitter.com/download/iphone" r...   
2335  <a href="http://twitter.com/download/iphone" r...   

                             

### ii. Accessing the tweet image prediction 

In [14]:
image_predictions_df.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [15]:
image_predictions_df.tail()

tweet_id                                          jpg_url  \
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num           p1   p1_conf  p1_dog                  p2   p2_conf  \
2070        2       basset  0.555712    True    English_springer  0.225770   
2071        1  paper_towel  0.170278   False  Labrador_retriever  0.168086   
2072        1    Chihuahua  0.716012    True            malamute  0.078253   
2073        1    Chihuahua  0.323581    True            Pekinese  0.090647   
2074        1       orange  0.097049   False               bagel  0.085851   

      p2_dog                           p3   p3_conf  p3_dog  
2070    True  German_short-haired_pointer  0.175219    True  
2071    True                      spatula  0.040836   False  
2072    True                       kelpie  0.031379    True  
2073    True                     papillon  0.068957    True  
2074   False                       banana  0.076110   False

In [16]:
image_predictions_df.sample(10)

tweet_id                                            jpg_url  \
582   678798276842360832    https://pbs.twimg.com/media/CWuTbAKUsAAvZHh.jpg   
1685  814530161257443328    https://pbs.twimg.com/media/C03K2-VWIAAK1iV.jpg   
626   680801747103793152    https://pbs.twimg.com/media/CXKxkseW8AAjAMY.jpg   
47    666817836334096384    https://pbs.twimg.com/media/CUEDSMEWEAAuXVZ.jpg   
971   706593038911545345    https://pbs.twimg.com/media/Cc5Snc7XIAAMidF.jpg   
1753  824663926340194305    https://pbs.twimg.com/media/C3HLd0HXUAAUI2b.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
966   706291001778950144    https://pbs.twimg.com/media/Cc0_2tXXEAA2iTY.jpg   
629   680913438424612864    https://pbs.twimg.com/media/CXMXKKHUMAA1QN3.jpg   
1402  768970937022709760  https://pbs.twimg.com/ext_tw_video_thumb/76896...   

      img_num                   p1   p1_conf  p1_dog                  p2  \
582         1             Airedale  0.583122    True       silky_terrier   
1685        1     miniature_poodle  0.626913    True          toy_poodle   
626         1                  pug  0.996720    True  Labrador_retriever   
47          1  miniature_schnauzer  0.496953    True  standard_schnauzer   
971         1          four-poster  0.696423   False               quilt   
1753        1       English_setter  0.526488    True    golden_retriever   
3           1  Rhodesian_ridgeback  0.408143    True             redbone   
966         1       Border_terrier  0.587101    True        bull_mastiff   
629         1           Pomeranian  0.615678    True    golden_retriever   
1402        1           Pomeranian  0.182358    True    golden_retriever   

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
582   0.129567    True             Lakeland_terrier  0.094727    True  
1685  0.265582    True  soft-coated_wheaten_terrier  0.041614    True  
626   0.001439    True    Staffordshire_bullterrier  0.000518    True  
47    0.285276    True              giant_schnauzer  0.073764    True  
971   0.189312   False                       pillow  0.029409   False  
1753  0.402815    True                 Irish_setter  0.034418    True  
3     0.360687    True           miniature_pinscher  0.222752    True  
966   0.164087    True    Staffordshire_bullterrier  0.105011    True  
629   0.126455    True                    Chihuahua  0.087184    True  
1402  0.110658    True                    mousetrap  0.086399   False

In [17]:
image_predictions_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [ ]:
# I'm merely interested as to why this image was classified as a paper towel.
image_predictions_df.loc[2071, 'jpg_url']

In [ ]:
from IPython.display import Image
Image(url="https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg")

### iii. Getting the tweet data from twitter API

In [ ]:
tweet_data.head()

In [ ]:
tweet_data.sample(10)

In [ ]:
tweet_data.info()

### Quality issues
1. Inaccurate tweet id (integer instead of string).

2. Invalid timestamp data type (String not DateTime)

3. The Twitter enhanced archive data contains 181 retweets.

4. There are 78 reply to tweet in the Twitter enhanced archive data

5. Incorrect dog names like None, a, an, & the.

6. Predicted photo data are not all complete (2075 instead of 2356)

7. Instead of using spaces, the picture names uses underscore in p1, p2 and p3 column.

8. Inconsistent title case for p name

### Tidiness issues
1. There are four columns of data for the dog stage

2. Despite being split into three different dataframes, all the data are related.

## Cleaning Data
In this section, clean **all** of the issues you documented while assessing. 

**Note:** Make a copy of the original data before cleaning. Cleaning includes merging individual pieces of data according to the rules of [tidy data](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html). The result should be a high-quality and tidy master pandas DataFrame (or DataFrames, if appropriate).

In [ ]:
# Make copies of original pieces of data
clean_twitter_archive = twitter_archive_df.copy()
clean_tweet_data = tweet_data.copy()
clean_image_predictions = image_predictions_df.copy()

In [ ]:
clean_twitter_archive.head()

In [ ]:
clean_tweet_data.head()

In [ ]:
clean_image_predictions.head()

In [ ]:
# Create a new dog_stage column by extracting dog stage from the text column.
clean_twitter_archive['dog_stage'] = clean_twitter_archive['text'].str.extract('(doggo|floofer|pupper|puppo)')

## Tidiness Issues

### **Issue #1:** There are four columns of data for the dog stage.

#### **Define:** The dog stage should be extracted from the text and merged into one colum (dog_stage), finally the four other colums should be dropped.

#### Code

In [ ]:
# Create a new dog_stage column by extracting dog stage from the text column.
clean_twitter_archive['dog_stage'] = clean_twitter_archive['text'].str.extract('(doggo|floofer|pupper|puppo)')
clean_twitter_archive = clean_twitter_archive.drop(columns=['doggo', 'floofer', 'pupper', 'puppo'])

#### Test

In [ ]:
clean_twitter_archive.sample(10)

### **Issue #2:** Despite being split into three different dataframes, all the data are related.

#### **Define:** All dataframes should be combined and joined together using the tweet id.

#### Code

In [ ]:
# Combining the Enhanced Twitter Archive data that has been cleaned with the Tweet Data from the Twitter API
clean_twitter_archive = pd.merge(clean_twitter_archive, clean_tweet_data, on='tweet_id', how='left')

# Taking the resulting merged archive and merging it with the Tweet Image Predictions
clean_twitter_archive = pd.merge(clean_twitter_archive, clean_image_predictions, on='tweet_id', how='left')

#### Test

In [ ]:
clean_twitter_archive.info()

## Quality Issues

### **Issue #1:** Inaccurate tweet id (integer instead of string).

#### **Define:** Convert the tweet id column's data type from an integer to a string using astype.

#### Code

In [ ]:
# Converting the tweet_id from integer into string
clean_twitter_archive.tweet_id = clean_twitter_archive.tweet_id.astype('str')

#### Test

In [ ]:
clean_twitter_archive.info()

### **Issue #2:** Invalid timestamp data type (String not DateTime).

#### **Define:** Convert the timestamp column's data type from an string to a DateTime.

#### Code

In [ ]:
# Converting the tweet_id from integer into string
clean_twitter_archive.timestamp = pd.to_datetime(clean_twitter_archive.timestamp)

#### Test

In [ ]:
clean_twitter_archive.info()

### **Issue #3:** The Twitter enhanced archive data contains 181 retweets.

#### **Define:** Remove all associated columns and the rows that indicate retweets.

#### Code

In [ ]:
# Only keep original tweets that do not have a retweet statud id.
clean_twitter_archive = clean_twitter_archive[clean_twitter_archive.retweeted_status_id.isnull()]
clean_twitter_archive.info()

In [ ]:
# Drop all retweet related columns
clean_twitter_archive = clean_twitter_archive.drop(columns=['retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp'])

#### Test

In [ ]:
clean_twitter_archive.info()

### **Issue #4:** There are 78 reply to tweet in the Twitter enhanced archive data.

#### **Define:** Remove all associated columns and the rows that indicate tweet replies.

#### Code

In [ ]:
# Only keep original tweets that do not have a retweet statud id.
clean_twitter_archive = clean_twitter_archive[clean_twitter_archive.in_reply_to_status_id.isnull()]
clean_twitter_archive.info()

In [ ]:
# Drop all retweet related columns
clean_twitter_archive = clean_twitter_archive.drop(columns=['in_reply_to_status_id', 'in_reply_to_user_id'])

#### Test

In [ ]:
clean_twitter_archive.info()

### **Issue #5:** Incorrect dog names like None, a, an, & the.

#### **Define:** Extract the correct names from the text column and change any invalid names (None or names beginning with lowercase letters) to NaN. (immediately following the word "named").

#### Code

In [ ]:
clean_twitter_archive.name = clean_twitter_archive.name.replace(regex=['^[a-z]+', 'None'], value= np.nan)

In [ ]:
# a function to extract names from text columns, returning NaN if there are no words that can be called names.
def name_extractor(text):
  list_of_text = text.split()
  for word in list_of_text:
    if word.lower() == 'named':
      name_index = list_of_text.index(word) + 1 # word after 'named
      return list_of_text[name_index]
    else:
      pass
  return np.nan

In [ ]:
#np.where(condition, what to do if condition is true, what to do if condition is false)
clean_twitter_archive.name = np.where(clean_twitter_archive.name.isnull(), clean_twitter_archive.text.apply(name_extractor), clean_twitter_archive.name)

#### Test

In [ ]:
clean_twitter_archive.info()

### **Issue #6:** Predicted photo data are not all complete (2075 instead of 2356).

#### **Define:** Remove the row with missing photo data.

#### Code

In [ ]:
clean_twitter_archive = clean_twitter_archive[clean_twitter_archive.jpg_url.notnull()]

#### Test

In [ ]:
clean_twitter_archive.info()

### **Issue #7:** Instead of using spaces, the picture names uses underscore in p1, p2 and p3 column.

#### **Define:** Replace the underscore with space.

#### Code

In [ ]:
clean_twitter_archive.p1 = clean_twitter_archive.p1.str.replace('_', ' ')
clean_twitter_archive.p2 = clean_twitter_archive.p2.str.replace('_', ' ')
clean_twitter_archive.p3 = clean_twitter_archive.p3.str.replace('_', ' ')

#### Test

In [ ]:
clean_twitter_archive.p1.sample(10)

In [ ]:
clean_twitter_archive.p2.sample(10)

In [ ]:
clean_twitter_archive.p3.sample(10)

### **Issue #8:** Inconsistent title case for p name.

#### **Define:** Replace the  names that starts with lowercase using an uppercase.

#### Code

In [ ]:
clean_twitter_archive.p1 = clean_twitter_archive.p1.str.title()
clean_twitter_archive.p2 = clean_twitter_archive.p2.str.title()
clean_twitter_archive.p3 = clean_twitter_archive.p3.str.title()

#### Test

In [ ]:
clean_twitter_archive.p1.sample(10)

In [ ]:
clean_twitter_archive.p2.sample(10)

In [ ]:
clean_twitter_archive.p3.sample(10)

In [ ]:
(clean_twitter_archive.p1.str.istitle()).value_counts()

In [ ]:
(clean_twitter_archive.p2.str.istitle()).value_counts()

In [ ]:
(clean_twitter_archive.p3.str.istitle()).value_counts()

## Storing Data
Save gathered, assessed, and cleaned master dataset to a CSV file named "twitter_archive_master.csv".

In [ ]:
clean_twitter_archive.to_csv('twitter_archive_master.csv')

## Analyzing and Visualizing Data
The data set is analysed in this section, along with the related visuals that help us reach insightful conclusions.

### Insights:
1. Displaying the general quantity of tweets over time allowing us to spot any noticeable changes in the volume of tweets.

2. The popularity of different dog stages in percentage.

3. Relationship between the number of retweets and the number of favorites.

### Visualization
1. Number of tweet as date changes

In [ ]:
clean_twitter_archive.timestamp = pd.to_datetime(clean_twitter_archive['timestamp'], format='%Y-%m-%d %H:%M:%S.%f')

monthly_tweets = clean_twitter_archive.groupby(pd.Grouper(key = 'timestamp', freq = "M")).count().reset_index()
monthly_tweets = monthly_tweets[['timestamp', 'tweet_id']]
monthly_tweets.head()
monthly_tweets.sum()

In [ ]:
plt.figure(figsize=(10, 8));
plt.xlim([datetime.date(2015, 11, 30), datetime.date(2017, 7, 30)]);

plt.xlabel('Year and Month')
plt.ylabel('Tweets Count')

plt.plot(monthly_tweets.timestamp, monthly_tweets.tweet_id);
plt.title('We Rate Dogs Tweets over Time');

Over time, the number of tweets drastically reduced, peaking in January and March of 2016 before steadily declining thereafter.

2. Percentage of dog stages

In [ ]:
dog_stage_df = clean_twitter_archive.dog_stage.value_counts()
dog_stage_df

In [ ]:
# Creating a pie chart
plt.pie(dog_stage_df,
        labels = ['Pupper', 'Doggo', 'Puppo', 'Floofer'], 
        autopct='%1.1f%%',
        shadow=True,
        explode=(0.1, 0.2, 0.2, 0.3)
        )

plt.title('Percentage of dog stages')
plt.axis('equal')

3. Relationship between the number of retweets and the number of favorites


In [ ]:
plt.scatter(clean_twitter_archive.retweet_count, clean_twitter_archive.favorite_count)
plt.title('Relationship between the number of retweets and the number of favorites')
plt.xlabel('Number of Retweets')
plt.ylabel ('Number of Favorites')